### ISSUE TITLE
Implement Lapidist YAML to determine which gems to release and for dependency tracking

### ISSUE THREAD

**Issue by opoudjis** [metanorma/lapidist#7]
I am getting errors in Travis metanorma-standoc which I am not getting in local compilation: the Travis compilation is ignoring post-release updates to relaton-iso-bib: https://github.com/relaton/relaton-iso-bib/releases

I note from the Gemfile.devel being run in lapidist the following:

````
gem 'isodoc', :github => '/isodoc', :branch => 'test/20190923'
gem 'relaton', :github => '/relaton', :branch => 'test/20190923'
gem 'html2doc', :github => '/html2doc', :branch => 'test/20190923'
gem 'metanorma', :github => '/metanorma', :branch => 'test/20190923'
````

This seems to me to indicate that the change in relaton-iso-bib is *not* being inherited through to the Travis run.

With these Gemfile.devel github pulls, they have to explore the inheritance trail exhaustively: if relaton is pulled, the *entire stack* of relaton needs to be pulled. The same applies for any other gems. This is the Gemfile.devel for metanorma-gb

````
gem 'metanorma-iso', :github => '/metanorma-iso', :branch => 'test/20190923'
gem 'isodoc', :github => '/isodoc', :branch => 'test/20190923'
gem 'metanorma', :github => '/metanorma', :branch => 'test/20190923'
````

metanorma-iso inherits from metanorma-standoc. Therefore, metanorma-standoc must also be pulled from github.

**Comment by ronaldtse** [comment]
@opoudjis @CAMOBAP795 This is a clear case for the `lapidist.yaml` file in order to configure these Gemfiles. We *should not* be using a `Gemfile.devel`.

Regarding auto-discovery, it is difficult to automatically find out all dependent gems because it is hard to know which gems are managed by us (i.e. part of the release) and which gems are external.

**Comment by CAMOBAP** [comment]
> This is a clear case for the lapidist.yaml file in order to configure these Gemfiles.

Custom config will not help with CI unless we decided to push our CI configuration on another level of complexity

> We should not be using a Gemfile.devel

This is a key thing, IDK how we can make CI works without 'overriding' gem sources

In this case, we have a problem that organization didn't get recognized, I will get invalid `Gemfile.devel`:
 - `gem 'metanorma', :github => '/metanorma', :branch => 'test/20190923'`
While it should be:
 - `gem 'metanorma', :github => 'metanorma/metanorma', :branch => 'test/20190923'`

Also, we pass gem list only ONCE during run `lapidist start ...` after this we refer to these gems via feature branch name 

> Regarding auto-discovery, it is difficult to automatically find out all dependent gems because it is hard to know which gems are managed by us (i.e. part of the release) and which gems are external.

I don't think so we can list of GitHub organization which is related to us, and treat all gems in those repositories as our gems

Proposed action items:
 - fix organization detection (I will work on this in scope of this task), afaik our gems can be only `relaton` or `metanorma`, right?

**Comment by ronaldtse** [comment]
@CAMOBAP795 I've added the `lapidist.yaml` files to:

* metanorma-release: https://github.com/metanorma/metanorma-release/commit/8a8a63305959dd76cddb991ef96e83b4c257f920
* relaton-release: https://github.com/relaton/relaton-release/commit/a82240c0803d69a45dfc582af99edf409c7fd35d

There are two things we have to distinguish:
* Gems that are part of the release (i.e. these gems are being released together)
* Gems that we need development versions but are NOT part of the release (these gems may or may not be managed by us)

I don't think auto-discovery of gem dependencies work that well because we need to distinguish the above two types of gems.

In our case, for example, in relaton-release, the relaton-release:lapidist.yaml file lists out all gems that are "part" of a Relaton release.

**Comment by ronaldtse** [comment]
The Gemfile used inside the to-be-released gems should be managed by Lapidist YAML, i.e. a line in `lapidist.yaml` that says:

```yaml
packages:
  unicode2latex:
     repo: https://github.com/metanorma/unicode2latex
  metanorma-standoc:
     repo: https://github.com/metanorma/metanorma-standoc
     dependencies:
       - html2doc
       - isodoc
       - iev
       - relaton
       - unicode2latex
```

Means that the `Gemfile` in `metanorma-standoc` should have:

```ruby
gem "unicode2latex", git: "https://github.com/metanorma/unicode2latex"
```

**Comment by CAMOBAP** [comment]
> I don't think auto-discovery of gem dependencies work that well because we need to distinguish the above two types of gems.

We list gems that are part of a feature only 2 times:
 - `lapidist start -b feature/X -g gem1,gem2,gem3`
 - `lapidist release -g gem1,gem2,gem3 --version minor`

We don't need to pass this list over and over again during using `lapidist` tool

BTW we can get repo URL by parsing `gem spec --remote metanorma` output, from repo url we can understand if this repo managed by us or not

We can distinguish them by organization whitelist for us to get repo name from dependency name we can parse 

> In our case, for example, in relaton-release, the relaton-release:lapidist.yaml file lists out all gems that are "part" of a Relaton release.

In my opinion, it will be 'hard' to maintain it, because we may forget to do keep `lapidist.yaml` up-to-date, in case of 

Also as I see `lapidist.yaml` cannot replace `Gemfile.devel` because we moved out of git submodules concept

**Comment by ronaldtse** [comment]
> We don't need to pass this list over and over again during using `lapidist` tool

We want to determine the full list of software within the "release" scope. Whether something is updated or not (i.e whether it needs to be version bumped, released or not) depends on the need  (if there is any change).

> We can distinguish them by organization whitelist for us to get repo name from dependency name we can parse

Not all gems that we own should belong to a release. For example, when we update metanorma, we do not want to update relaton.

> In my opinion, it will be 'hard' to maintain it, because we may forget to do keep lapidist.yaml up-to-date, in case of

I don't see why it is hard to keep lapidist.yaml up to date.

> Also as I see lapidist.yaml cannot replace Gemfile.devel because we moved out of git submodules concept

Whether we use git submodule or not is not important. The point is:

1. We need a full development tree with all the dependencies cross-linked. 
2. There is a command that runs the integration tests across these dependencies.
3. If there are changes detected, the changed software / gems need to be fixed 
4. Once all the tests pass, the changed software / gems need to be released. (Note that some of the gems do not need to be changed/released)

Number 1 can be achieved with individual git clone folders or git submodules.
